### PR TITLE
Battle tower

### PR DESCRIPTION
**Pull request by PikalaxALT**

### CODE REVIEW

File: src/battle_tower.c [hunk 1]
```diff
@@ -1,9 +1,46 @@
...
+#include "item.h"
+#include "script_pokemon_util_80A0058.h"
 #include "constants/species.h"
 #include "constants/items.h"
 #include "constants/moves.h"
 #include "constants/pokemon.h"
+#include "constants/vars.h"
+#include "constants/map_objects.h"
+
+EWRAM_DATA u16 gUnknown_203AAB8 = 0;
```
Comment (jiangzhengwenjz) [inline src/battle_tower.c]: Various functions and data should be static

File: src/battle_tower.c [hunk 2]
```diff
@@ -131,3 +168,1623 @@ const u16 sLongStreakPrizes[] = {
...
+{
+    s32 i;
+    u8 battleTowerLevelType = gSaveBlock2Ptr->battleTower.battleTowerLevelType;
+
+    switch (gSpecialVar_0x8004)
+    {
+    case 0:
+        // Weird cast required to match
+        // Leftover from RS macro?
+        *((u8 *)&gBattleStruct->arenaTurnCounter) = gSaveBlock2Ptr->battleTower.var_4AE[battleTowerLevelType];
```
Comment (jiangzhengwenjz) [inline src/battle_tower.c]: Can you rename this field?

Comment (PikalaxALT) [reply src/battle_tower.c]: It's `ewram160FB` in Ruby

File: src/battle_tower.c [hunk 3]
```diff
@@ -1,13 +1,51 @@
...
+static void SaveCurrentWinStreak(void);
+static void ValidateBattleTowerRecordChecksums(void);
+static void SetBattleTowerRecordChecksum(struct BattleTowerRecord * record);
+static void ClearBattleTowerRecord(struct BattleTowerRecord * record);
+static void PopulateBravoTrainerBattleTowerLostData(void);
+static u16 GetCurrentBattleTowerWinStreak(u8 levelType);
+static void SetEReaderTrainerChecksum(struct BattleTowerEReaderTrainer * eReaderTrainer);
+static void PrintEReaderTrainerFarewellMessage(void);
 
 #include "data/battle_tower/trainers.h"
```
Comment (jiangzhengwenjz) [inline src/battle_tower.c]: Aren't the arrays in the 3 included headers static?

Comment (PikalaxALT) [reply src/battle_tower.c]: Not in Ruby

**Pull request merged**